source/_ext/cookbook/proc_examples.py

```diff
@@ -0,0 +1,261 @@
+    # TODO: Work out how to store widget state for nglview
+    executor = ExecutePreprocessor(
+        kernel_name="python3",
+        timeout=600,
+    )
+    executor.store_widget_state = True
+    # Execute the notebook
+    # TODO: Do this in a temporary directory with needed_files to keep src clean
+    # TODO: Run in the notebook-specific Conda environment
+    executor.preprocess(nb, {"metadata": {"path": src.parent}})
+
+    # Write the executed notebook
+    dst = EXEC_IPYNB_ROOT / src.relative_to(SRC_IPYNB_ROOT)
+    dst.parent.mkdir(parents=True, exist_ok=True)
+    with open(dst, "w", encoding="utf-8") as f:
+        nbformat.write(nb, f)
+
+    print("Done executing", src.relative_to(SRC_IPYNB_ROOT))
+
+
+def delay_iterator(iterator, seconds=1.0):
+    """Introduce a delay to iteration.
+
+    This may be helpful to avoid race conditions with process pools."""
+    for item in iterator:
+        yield item
+        sleep(seconds)
+
+
+def main(do_proc=True, do_exec=True, redo_all=False):
+    print("Working in", Path().resolve())
+
+    # Download the examples from GitHub
+    updates = None
+    for repo in GITHUB_REPOS:
+        print("Downloading", repo, "to", (SRC_IPYNB_ROOT / repo).resolve())
+        updates = download_dir(repo, "examples", SRC_IPYNB_ROOT)
+    if updates is None:
+        print("Nothing to do.")
+        return
+
+    # Find the notebooks we need to update
+    notebooks = [*find_notebooks(SRC_IPYNB_ROOT)]
+    if updates.start_over or redo_all:
+        # If we're starting over, we do all the notebooks and clear what's
+        # already in there
+        print("Updating all notebooks")
+        shutil.rmtree(EXEC_IPYNB_ROOT, ignore_errors=True)
+        shutil.rmtree(COLAB_IPYNB_ROOT, ignore_errors=True)
+        shutil.rmtree(ZIPPED_IPYNB_ROOT, ignore_errors=True)
+    else:
+        # Only work on notebooks that were changed AND were found
+        # This is important because `updates` doesn't do any of the filtering
+        # that find_notebooks does (eg, deprecated notebooks)
+        notebooks = [
+            *set(notebooks).intersection(
+                [SRC_IPYNB_ROOT / path for path in updates.reprocess]
+            )
+        ]
+        # TODO: Clean up notebooks that have been removed
+
+    # Create Colab and downloadable versions of the notebooks
+    if do_proc:
+        for notebook in notebooks:
+            print("Processing", notebook)
+            create_colab_notebook(notebook)
+            create_zip(notebook)
+
+    # Execute notebooks in parallel for rendering as HTML
+    if do_exec:
+        # Context manager ensures the pool is correctly terminated if there's
+        # an exception
+        with Pool() as pool:
+            # Wait a second between launching subprocesses
+            # Workaround https://github.com/jupyter/nbconvert/issues/1066
+            _ = [*pool.imap_unordered(execute_notebook, delay_iterator(notebooks))]
```

Comment:
Nice, I've always wanted to be able to do this in tests but I couldn't figure out how to get nbval to do it. I mean, I [guess it's possible](https://github.com/computationalmodelling/nbval/blob/d7bc34893a9251fbfd4b270347197619c32fe6ea/README.md?plain=1#L114) and I just never found out how. Still have to turn off `pytest-randomly`, probably